File: bytopic/twopointers/java/group2/3Sum.java
// Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k,
// and nums[i] + nums[j] + nums[k] == 0.
// Notice that the solution set must not contain duplicate triplets.
// Example:
// Input: nums = [-1,0,1,2,-1,-4]
// Output: [[-1,-1,2],[-1,0,1]]
// Explanation:
// nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
// nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
// nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
// The distinct triplets are [-1,0,1] and [-1,-1,2].
// Notice that the order of the output and the order of the triplets does not matter.
// Tag: 29/150

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.HashSet;
import java.util.Set;
class ThreeSum {
    public static void main(String...args) {
        int[] nums = {-1,0,1,2,-1,-4};
        List<List<Integer>> result = threeSum(nums);
        System.out.println(result);
    }

    static List<List<Integer>> threeSum(int[] nums) {
        int target = 0;
        Arrays.sort(nums);
        Set<List<Integer>> s = new HashSet<>();
        List<List<Integer>> output = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            int j = i + 1;
            int k = nums.length - 1;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum == target) {
                    s.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++;
                    k--;
                } else if (sum < target) {
                    j++;
                } else {
                    k--;
                }
            }
        }
        output.addAll(s);
        return output;
    }
}

// Steps:
// 1. Sort the input array
// 2. Initialize a set to store the unique triplets and an output vector to store the final result
// 3. Iterate through the array with a variable i, starting from index 0.
// 4. Initialize two pointers, j and k, with j starting at i+1 and k starting at the end of the array.
// 5. In the while loop, check if the sum of nums[i], nums[j], and nums[k] is equal to 0. If it is, insert the triplet into the set
// and increment j and decrement k to move the pointers.
// 6. If the sum is less than 0, increment j. If the sum is greater than 0, decrement k.
// 7. After the while loop, iterate through the set and add each triplet to the output vector.
// 8. Return the output vector / list.

File: bytopic/twopointers/python/group2/3sum.py
"""
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k,
and nums[i] + nums[j] + nums[k] == 0.
Notice that the solution set must not contain duplicate triplets.

Example:
Input: nums = [-1,0,1,2,-1,-4]
Output: [[-1,-1,2],[-1,0,1]]
Explanation:
nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
The distinct triplets are [-1,0,1] and [-1,-1,2].
Notice that the order of the output and the order of the triplets does not matter.

Tag: 29/150
Tag: 15/2927, R15/2936 (overall frequency ranking)
"""


def three_sum(nums: list[int]) -> list[list[int]]:
    res = set()
    # 1. Split nums into three lists: negative numbers, positive numbers, and zeros
    n, p, z = [], [], []
    for num in nums:
        if num > 0:
            p.append(num)
        elif num < 0:
            n.append(num)
        else:
            z.append(num)
    # 2. Create a separate set for negatives and positives for O(1) look-up times
    set_n, set_p = set(n), set(p)
    # 3. If there is at least one zero in the list, all all cases where -num exists in set_n and num exists in set_p
    # e. g. (-3, 0, 3) = 0
    if z:
        for num in set_p:
            if -1 * num in set_n:
                res.add((-1 * num, 0, num))
    # 3. If there are at least 3 zeroes in the list then also include (0, 0, 0) = 0
    if len(z) >= 3:
        res.add((0, 0, 0))
    # 4. For all pairs of negative numbers (-3, -1), check to see if their complement (4) exists
    # in the positive number set
    for i in range(len(n)):
        for j in range(i+1, len(n)):
            target = -1 * (n[i] + n[j])
            if target in set_p:
                res.add(tuple(sorted([n[i], n[j], target])))
    # 5. For all pairs of positive numbers (1, 1), check to see if their complement (-2) exists in
    # the negative number set
    for i in range(len(p)):
        for j in range(i+1, len(p)):
            target = -1 * (p[i] + p[j])
            if target in set_n:
                res.add(tuple(sorted([p[i], p[j], target])))
    return res


def main():
    nums = [-1, 0, 1, 2, -1, -4]
    print(three_sum(nums))


if __name__ == "__main__":
    main()
